usr/src/uts/common/io/scsi/adapters/smartpqi/smartpqi_util.c

```diff
@@ -0,0 +1,947 @@
+		}
+
+		if (grab_lock == B_TRUE)
+			mutex_exit(&devp->pd_mutex);
+		break;
+
+	default:
+		/*
+		 * Normally a panic or ASSERT(0) would be called
+		 * for here. Except that in this case the 'cmd'
+		 * memory could be coming from the kmem_cache pool
+		 * which during debug gets wiped with 0xbaddcafe
+		 */
+		break;
+	}
+}
+
+
+static uint_t supported_event_types[] = {
+	PQI_EVENT_TYPE_HOTPLUG,
+	PQI_EVENT_TYPE_HARDWARE,
+	PQI_EVENT_TYPE_PHYSICAL_DEVICE,
+	PQI_EVENT_TYPE_LOGICAL_DEVICE,
+	PQI_EVENT_TYPE_AIO_STATE_CHANGE,
+	PQI_EVENT_TYPE_AIO_CONFIG_CHANGE,
+	PQI_EVENT_TYPE_HEARTBEAT
+};
+
+int
+pqi_map_event(uint8_t event)
+{
+	int i;
+
+	for (i = 0; i < sizeof (supported_event_types) / sizeof (uint_t); i++)
+		if (supported_event_types[i] == event)
+			return (i);
+	return (-1);
+}
+
+boolean_t
+pqi_supported_event(uint8_t event)
+{
+	return (pqi_map_event(event) == -1 ? B_FALSE : B_TRUE);
+}
+
+char *
+pqi_event_to_str(uint8_t event)
+{
+	switch (event) {
+	case PQI_EVENT_TYPE_HOTPLUG: return ("Hotplug");
+	case PQI_EVENT_TYPE_HARDWARE: return ("Hardware");
+	case PQI_EVENT_TYPE_PHYSICAL_DEVICE:
+		return ("Physical Device");
+	case PQI_EVENT_TYPE_LOGICAL_DEVICE: return ("logical Device");
+	case PQI_EVENT_TYPE_AIO_STATE_CHANGE:
+		return ("AIO State Change");
+	case PQI_EVENT_TYPE_AIO_CONFIG_CHANGE:
+		return ("AIO Config Change");
+	case PQI_EVENT_TYPE_HEARTBEAT: return ("Heartbeat");
+	default: return ("Unsupported Event Type");
+	}
+}
+
+char *
+bool_to_str(int v)
+{
+	return (v ? "T" : "f");
+}
+
+char *
+dtype_to_str(int t)
+{
+	switch (t) {
+	case DTYPE_DIRECT: return ("Direct");
+	case DTYPE_SEQUENTIAL: return ("Sequential");
+	case DTYPE_ESI: return ("ESI");
+	case DTYPE_ARRAY_CTRL: return ("RAID");
+	default: return ("Ughknown");
+	}
+}
+
+static ddi_dma_attr_t single_dma_attrs = {
+	DMA_ATTR_V0,	/* attribute layout version			*/
+	0x0ull,		/* address low - should be 0 (longlong)		*/
+	0xffffffffffffffffull, /* address high - 64-bit max		*/
+	0x7ffffull,	/* count max - max DMA object size		*/
+	4096,		/* allocation alignment requirements		*/
+	0x78,		/* burstsizes - binary encoded values		*/
+	1,		/* minxfer - gran. of DMA engine		*/
+	0x007ffffull,	/* maxxfer - gran. of DMA engine		*/
+	0xffffffffull,	/* max segment size (DMA boundary)		*/
+	1,		/* For pqi_alloc_single must be contig memory	*/
+	512,		/* granularity - device transfer size		*/
+	0		/* flags, set to 0				*/
+};
+
+pqi_dma_overhead_t *
+pqi_alloc_single(pqi_state_t s, size_t len)
+{
+	pqi_dma_overhead_t	*d;
+	ddi_dma_cookie_t	cookie;
+
+	d = kmem_zalloc(sizeof (*d), KM_SLEEP);
+	d->len_to_alloc = len;
+
+	if (ddi_dma_alloc_handle(s->s_dip, &single_dma_attrs,
+	    DDI_DMA_SLEEP, 0, &d->handle) != DDI_SUCCESS)
+		goto error_out;
+
+	if (ddi_dma_mem_alloc(d->handle, len, &s->s_reg_acc_attr,
+	    DDI_DMA_CONSISTENT, DDI_DMA_SLEEP, 0,
+	    &d->alloc_memory, &len, &d->acc) != DDI_SUCCESS)
+		goto error_out;
+
+	(void) memset(d->alloc_memory, 0, len);
```

Comment:
In kernel space, bzero() is preferred.  (It's actually very slightly more efficient.)